File: Elixir/lib/War.ex
defmodule War do
  
  @moduledoc """    
	The `War` module applies function “overloading” (pattern matching on the signature) 
	on the function deal. The deal function will generate two players' piles from the 
	already-shuffled deck. When one player wins the round, either instantly or after war(s), 
	cards on the warchest will be added to the bottom of that player's deck. These cards 
	should be added in decreasing order by rank, i.e., first place the highest ranked card 
	on the bottom, then place the next highest ranked card beneath that. Ace, denoted as 1, 
	has the highest rank, Two, denoted as 2, has the lowest. The deal function will return 
	the pile of the winner. This pile will contain all 52 integers from the original input 
	permutation and be in the correct order, i.e, pass all 21 test cases.
	"""


  @doc """
	Function stub for deal/1 is given below. Feel free to add 
    as many additional helper functions as you want. 
    The tests for the deal function can be found in test/war_test.exs. 
    You can add your five test cases to this file.
    Run the tester by executing 'mix test' from the war directory 
    (the one containing mix.exs)
	"""


  # deal/1 is the main function receiving the original permutation.
  # deal/1 generates 2 piles, a1 and a2, and starts the game.
  # The order of a1 and a2 is that the top card of each deck will be the first element in list.
  def deal(shuf) do
	p1 = List.flatten(Enum.chunk_every(shuf,1,2))
	p2 = List.flatten(Enum.chunk_every(tl(shuf),1,2))
	a1 = Enum.reverse(Enum.map(p1, fn x -> if x == 1, do: 14, else: x end))
	a2 = Enum.reverse(Enum.map(p2, fn x -> if x == 1, do: 14, else: x end))
	Enum.map(deal(a1, a2, []), fn x -> if x == 14, do: 1, else: x end)
  end
  
  
  #deal/3 manages the scenario when player 1 has no cards left.
  #Player 2 wins and the pile onHold will be added to the bottom of player 2's pile.
  #The scenario that p2 may also be empty can also be handled in this function.
  #The winner's pile will be returned.
  def deal([], p2, onHold) do
	p2 ++ onHold
  end
  
  
  #deal/3 manages the scenario when player 2 has no cards left.
  #Player 1 wins and the pile onHold will be added to the bottom of player 1's pile.
  #The scenario that p1 may also be empty can also be handled in this function.
  #The winner's pile will be returned.  
  def deal(p1, [], onHold) do
	p1 ++ onHold
  end
  
  
  #deal/3 manages the scenario when both players have non-empty decks.
  def deal([h1|t1], [h2|t2], onHold) do
	onHold = Enum.sort(onHold ++ [h2] ++ [h1], :desc)
	cond do
		#Non-war situation
		h1 < h2 ->
			deal(t1, t2 ++ onHold, [])
		
		#Non-war situation
		h2 < h1 ->
			deal(t1 ++ onHold, t2, [])
		
		#War!!!	
		h1 == h2 ->
			cond do	
				#Handle empty scenario with grace
				Enum.empty?(t1) or Enum.empty?(t2) ->
					deal(t1, t2, onHold)
				
				#Both decks are non-empty
				true ->
					onHold = Enum.sort(onHold ++ [hd(t1)] ++ [hd(t2)], :desc)
					deal(tl(t1), tl(t2), onHold)
			end
	end
  
  end

end

File: Haskell/lib/War.hs
module War (deal) where

import Data.List

{--
Function stub(s) with type signatures for you to fill in are given below. 
Feel free to add as many additional helper functions as you want. 

The tests for these functions can be found in src/TestSuite.hs. 
You are encouraged to add your own tests in addition to those provided.

Run the tester by executing 'cabal test' from the war directory 
(the one containing war.cabal)
--}


-- deal is the main function receiving the original permutation.
-- deal will call startGame to generate 2 piles, p1 and p2, and let the game begin!
-- The order of p1 and p2 is that the top card of each deck will be the first element in list.
deal :: [Int] -> [Int]
deal shuf = do
 let [p1,p2] = startGame (convertTo14 (shuf, []), [], [])
 myDeal (p1, p2, [])

-- convertTo14 converts number '1', to '14', so that Ace will be the greatest number in the list.
convertTo14 (shuf, res)
 | length shuf == 0 = res
 | head shuf == 1 = convertTo14 (tail shuf, res ++ [14])
 | otherwise = convertTo14 (tail shuf, res ++ [head shuf])

-- After the game comes to an end, convertToAce will recover the correct denomination of Ace, changing 14 to 1.
convertToAce (shuf, res)
 | length shuf == 0 = res
 | head shuf == 14 = convertToAce (tail shuf, res ++ [1])
 | otherwise = convertToAce (tail shuf, res ++ [head shuf])

-- myDeal is the function that implements the game
myDeal (p1, p2, onHold)
 | length p1 == 0 = convertToAce (p2 ++ onHold, [])
 | length p2 == 0 = convertToAce (p1 ++ onHold, [])
 | head p1 == head p2 = warNow (tail p1, tail p2, reverse (sort (onHold ++ [head p1, head p2])))
 | head p1 < head p2 = myDeal (tail p1, tail p2 ++ reverse (sort (onHold ++ [head p1, head p2])), [])
 | head p2 < head p1 = myDeal (tail p1 ++ reverse (sort (onHold ++ [head p1, head p2])), tail p2, [])

-- warNow is the function that deals with War scenarios, i.e, face-up cards are equal.
warNow (p1, p2, onHold)
 | length p1 == 0 = convertToAce (p2 ++ onHold, [])
 | length p2 == 0 = convertToAce (p1 ++ onHold, [])
 | otherwise = myDeal (tail p1, tail p2, reverse (sort (onHold ++ [head p1, head p2])))

-- startGame generates 2 players' piles and let the game begin!
startGame (shuf, p1, p2)
 | length shuf == 0 = [p1, p2]
 | otherwise = startGame (tail (tail shuf), (head shuf) : p1, head (tail shuf) : p2)
 